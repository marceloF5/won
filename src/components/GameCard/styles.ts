import styled, { css, DefaultTheme } from 'styled-components'

type PriceProps = {
    isPromotional?: boolean
}

const priceModifiers = {
    default: (theme: DefaultTheme) => css`
        background-color: ${theme.colors.secondary};
        border-radius: ${theme.border.radius};
        color: ${theme.colors.white};
        margin-right: calc(${theme.spacings.xxsmall} / 2);
        padding: 0 ${theme.spacings.xxsmall};
    `,

    promotional: (theme: DefaultTheme) => css`
        color: ${theme.colors.gray};
        margin-right: ${theme.spacings.xsmall};
        text-decoration: line-through;
    `
}

export const Wrapper = styled.article`
    ${({ theme }) => css`
        background-color: ${theme.colors.white};
        height: 100%;
        width: 100%;

        position: relative;
        display: flex;
        flex-direction: column;
    `}
`

export const ImageBox = styled.a`
    height: 14rem;
    width: 100%;
    background: #f6f7f8;
    background-image: linear-gradient(
        to right,
        #f6f7f8 0%,
        #edeef1 20%,
        #f6f7f8 40%,
        #f6f7f8 100%
    );
    background-size: 80rem 14rem;
    animation: placeholderShimmer 1s linear infinite forwards;

    img {
        width: 100%;
        height: 100%;
        object-fit: cover;
    }

    @keyframes placeholderShimmer {
        0% {
            background-position: -40rem 0;
        }

        100% {
            background-position: 40rem 0;
        }
    }
`

export const Content = styled.div`
    ${({ theme }) => css`
        height: 100%;
        margin: ${theme.spacings.xsmall};

        display: flex;
        flex-direction: column;
        justify-content: space-between;
        position: relative;
    `}
`

export const Info = styled.a`
    max-width: calc(100% - 2.5rem);
    text-decoration: none;
`

export const Title = styled.h3`
    ${({ theme }) => css`
        color: ${theme.colors.black};
        font-size: ${theme.font.sizes.medium};
        font-weight: ${theme.font.bold};
        line-height: ${theme.font.sizes.medium};
    `}
`

export const Developer = styled.h4`
    ${({ theme }) => css`
        color: ${theme.colors.gray};
        font-size: ${theme.font.sizes.small};
        font-weight: ${theme.font.bold};
    `}
`

export const FavButton = styled.div`
    ${({ theme }) => css`
        color: ${theme.colors.primary};
        position: absolute;
        right: 0;
        top: -0.5rem;
        cursor: pointer;

        svg {
            width: 2.5rem;
        }
    `}
`

export const BuyBox = styled.div`
    ${({ theme }) => css`
        margin-top: ${theme.spacings.xxsmall};

        display: flex;
        align-items: center;
        justify-content: flex-end;
    `}
`

export const Price = styled.div<PriceProps>`
    ${({ theme, isPromotional }) => css`
        font-weight: ${theme.font.bold};
        height: 3rem;

        display: inline-flex;
        align-items: center;
        ${!isPromotional && priceModifiers.default(theme)}
        ${isPromotional && priceModifiers.promotional(theme)}
    `}
`
