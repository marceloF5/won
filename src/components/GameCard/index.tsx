import {
    Favorite,
    FavoriteBorder,
    AddShoppingCart
} from '@styled-icons/material-outlined'

import Button from 'components/Button'
import Ribbon, { RibbonColors, RibbonSizes } from 'components/Ribbon'
import * as S from './styles'

export type GameCardProps = {
    title: string
    developer: string
    img: string
    price: string
    favorite?: boolean
    promotionalPrice?: string
    ribbon?: React.ReactNode
    ribbonColor?: RibbonColors
    ribbonSize?: RibbonSizes
    onFav?: () => void
}

const GameCard = ({
    title,
    developer,
    img,
    price,
    favorite = false,
    promotionalPrice,
    ribbon,
    ribbonColor,
    ribbonSize,
    onFav
}: GameCardProps) => (
    <S.Wrapper>
        {!!ribbon && (
            <Ribbon color={ribbonColor} size={ribbonSize}>
                {ribbon}
            </Ribbon>
        )}
        <S.ImageBox>
            <img src={img} alt={title} />
        </S.ImageBox>
        <S.Content>
            <S.Info>
                <S.Title>{title}</S.Title>
                <S.Developer>{developer}</S.Developer>
            </S.Info>
            <S.FavButton onClick={onFav} role="button">
                {favorite ? (
                    <Favorite aria-label="Remove from Wishlist" />
                ) : (
                    <FavoriteBorder aria-label="Add to Wishlist" />
                )}
            </S.FavButton>
            <S.BuyBox>
                {!!promotionalPrice && <S.Price isPromotional>{price}</S.Price>}
                <S.Price>{promotionalPrice || price}</S.Price>
                <Button icon={<AddShoppingCart />} size="small" />
            </S.BuyBox>
        </S.Content>
    </S.Wrapper>
)

export default GameCard
