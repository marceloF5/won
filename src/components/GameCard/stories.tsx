import { Story, Meta } from '@storybook/react/types-6-0'
import GameCard, { GameCardProps } from '.'

export default {
    title: 'GameCard',
    component: GameCard,
    args: {
        title: 'Population Zero',
        developer: 'Rockestar Games',
        img: 'https://source.unsplash.com/user/willianjusten/1042x580',
        price: '€ 65.00'
    },
    argTypes: {
        onFav: { action: 'clicked' },
        ribbon: { type: 'string' }
    },
    parameters: {
        backgrounds: {
            default: 'dark'
        }
    }
} as Meta

export const Default: Story<GameCardProps> = (args) => (
    <div style={{ width: '30rem' }}>
        <GameCard {...args} />
    </div>
)

export const WithRibbon: Story<GameCardProps> = (args) => (
    <div style={{ width: '30rem' }}>
        <GameCard {...args} />
    </div>
)

WithRibbon.args = {
    promotionalPrice: '€ 20.00',
    ribbon: '80% OFF',
    ribbonSize: 'small',
    ribbonColor: 'primary'
}
