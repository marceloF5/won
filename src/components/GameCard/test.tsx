import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils/tests/helpers'
import GameCard from '.'

const props = {
    title: 'Population Zero',
    developer: 'Rockestar Games',
    img: 'https://source.unsplash.com/user/willianjusten/1042x580',
    price: '€ 65.00'
}

describe('<GameCard />', () => {
    it('should render correctly', () => {
        renderWithTheme(<GameCard {...props} />)

        expect(
            screen.getByRole('heading', { name: props.title })
        ).toBeInTheDocument()
        expect(
            screen.getByRole('heading', { name: props.developer })
        ).toBeInTheDocument()
        expect(screen.getByRole('img', { name: props.title })).toHaveAttribute(
            'src',
            props.img
        )

        expect(screen.getByLabelText(/add to wishlist/i)).toBeInTheDocument()
    })

    it('should render price in label', () => {
        renderWithTheme(<GameCard {...props} />)

        const price = screen.getByText('€ 65.00')

        expect(price).not.toHaveStyle({ textDecoration: 'line-through' })
        expect(price).toHaveStyle({
            backgroundColor: '#3cd3c1'
        })
    })

    it('should render a line-through in price when promotional', () => {
        renderWithTheme(<GameCard {...props} promotionalPrice="€ 20.00" />)

        const promotionalPrice = screen.getByText('€ 65.00')
        const price = screen.getByText('€ 20.00')

        expect(promotionalPrice).toHaveStyle({ textDecoration: 'line-through' })
        expect(price).not.toHaveStyle({ textDecoration: 'line-through' })
    })

    it('should render a fulled Favorite icon when favorite is true', () => {
        renderWithTheme(<GameCard {...props} favorite />)

        expect(
            screen.getByLabelText(/remove from wishlist/i)
        ).toBeInTheDocument()
    })

    it('should call on onFav method favorite is clicked', () => {
        const onFav = jest.fn()

        renderWithTheme(<GameCard {...props} favorite onFav={onFav} />)

        const favButton = screen.getAllByRole('button')[0]

        fireEvent.click(favButton)
    })
})
