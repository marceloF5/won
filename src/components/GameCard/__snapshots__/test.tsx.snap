// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameCard /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    GameCard
  </h1>
</main>
`
